core/src/main/java/org/bitcoinj/core/Wallet.java

```diff
@@ -4012,12 +4146,44 @@ public void reorganize(StoredBlock splitPoint, List<StoredBlock> oldBlocks, List
-                confidenceChanged.put(tx, TransactionConfidence.Listener.ChangeReason.TYPE);
-                addWalletTransaction(Pool.PENDING, tx);
-                updateForSpends(tx, false);
+                
+
+                Set<Transaction> doubleSpendTxns = findDoubleSpendsAgainst(tx, dead);
+                if (doubleSpendTxns.isEmpty()) {
+                    // No double spends detected.
+                    // Update the confidence type of tx to PENDING and schedule confidence listener notifications.                     
+                    if (!tx.getConfidence().getConfidenceType().equals(ConfidenceType.IN_CONFLICT)) {
+                        tx.getConfidence().setConfidenceType(ConfidenceType.PENDING);                
+                        confidenceChanged.put(tx, TransactionConfidence.Listener.ChangeReason.TYPE);
+                        addWalletTransaction(Pool.PENDING, tx);           
+                        updateForSpends(tx, false);                        
+                    }
+                } else {
+                    // Double spends detected.
+                    // Update the confidence type of the supplied tx and the other txns in conflict to IN_CONFLICT,
+                    // move the other txns from the dead pool to the pending pool and schedule confidence 
+                    // listener notifications.                      
+                    doubleSpendTxns.add(tx);
+                    Set<Transaction> txPool = Sets.newHashSet(getTransactions(true));
+                    txPool.addAll(oldChainTxns);
+                    addTransactionsDependingOn(doubleSpendTxns, txPool);
+                    List<Transaction> doubleSpendTxnsSorted = sortTxnsByDependency(doubleSpendTxns);
+                    for (Transaction doubleSpendTx : doubleSpendTxnsSorted) {
+                        if (!doubleSpendTx.getConfidence().getConfidenceType().equals(ConfidenceType.IN_CONFLICT)) {
+                            if (isNotSpendingTxnsInConfidenceType(doubleSpendTx, ConfidenceType.DEAD)) {
+                                dead.remove(doubleSpendTx.getHash());                        
+                                pending.remove(doubleSpendTx.getHash());                        
+                                addWalletTransaction(Pool.PENDING, doubleSpendTx);
+                                doubleSpendTx.getConfidence().setConfidenceType(ConfidenceType.IN_CONFLICT);                    
+                                confidenceChanged.put(doubleSpendTx, TransactionConfidence.Listener.ChangeReason.TYPE);
+                                updateForSpends(doubleSpendTx, false);                                
+                            }
+                        }                    
+                    }
+                }        
             }
 
+            // Note: Is this comment still valid after we implement support for double spend forwarding? I guess it is not.
```

Comment:
I think it is. The first double spend will be forwarded *if* the user is connected to Bitcoin XT nodes, but the behaviour on re-org is the same: double spent txns are erased from the mempool and will not be resurrected by miners.